maintainers_annotation/maintainers_annotation.go

```diff
@@ -0,0 +1,190 @@
+	AnnotationNotApplicable: "The check or probe is not applicable in this case.",
+	AnnotationNotSupported:  "The check or probe is fulfilled but in a way that is not supported by Scorecard.",
+	AnnotationNotDetected:   "The check or probe is fulfilled but in a way that is supported by Scorecard but it was not detected.",
+}
+
+// Annotation groups the annotation reason and, in the future, the related probe.
+// If there is a probe, the reason applies to the probe.
+// If there is not a probe, the reason applies to the check or group of checks in
+// the exemption.
+type Annotation struct {
+	AnnotationReason AnnotationReason `yaml:"annotation"`
+}
+
+// Exemption defines a group of checks that are being annotated for various reasons.
+type Exemption struct {
+	Checks      []string     `yaml:"checks"`
+	Annotations []Annotation `yaml:"annotations"`
+}
+
+// MaintainersAnnotation is a group of exemptions.
+type MaintainersAnnotation struct {
+	Exemptions []Exemption `yaml:"exemptions"`
+}
+
+// parseScorecardYmlFile takes the scorecard.yml file content and returns a `MaintainersAnnotation`.
+func parseScorecardYmlFile(ma *MaintainersAnnotation, content []byte) error {
+	unmarshalErr := yaml.Unmarshal(content, ma)
+	if unmarshalErr != nil {
+		return sce.WithMessage(sce.ErrScorecardInternal, unmarshalErr.Error())
+	}
+
+	return nil
+}
+
+// validateAndParseScorecardYmlFile takes the scorecard.yml files found, validates if its the correct
+// configuration file and parse its configurations.
+func validateAndParseScorecardYmlFile(path string, content []byte, args ...interface{}) (bool, error) {
+	// Verify if received the necessary parameters to save the scorecard YML configurations
+	if len(args) != 1 {
+		return false, fmt.Errorf(
+			"validateAndParseScorecardYmlFile requires exactly one argument to save maintainers annotation data")
+	}
+	ma, ok := args[0].(*MaintainersAnnotation)
+	if !ok {
+		return false, fmt.Errorf(
+			"validateAndParseScorecardYmlFile requires argument of type *MaintainersAnnotation")
+	}
+
+	// Exclude scorecard GitHub Action file
+	if strings.Contains(path, ".github") {
+		return true, nil
+	}
+
+	// Keep looking for files if content is empty
+	if len(content) == 0 {
+		return true, nil
+	}
+
+	// Found scorecard.yml file, parse file
+	err := parseScorecardYmlFile(ma, content)
+	if err != nil {
+		return false, err
+	}
+
+	return false, nil
+}
+
+// readScorecardYmlFromRepo reads the scorecard.yml file from the repo and returns its configurations.
+func readScorecardYmlFromRepo(repoClient clients.RepoClient) (MaintainersAnnotation, error) {
+	ma := MaintainersAnnotation{}
+	err := fileparser.OnMatchingFileContentDo(repoClient, fileparser.PathMatcher{
+		Pattern:       "scorecard.yml",
+		CaseSensitive: false,
+	}, validateAndParseScorecardYmlFile, &ma)
+
+	if err != nil {
+		return ma, fmt.Errorf("%w", err)
+	}
+
+	if ma.Exemptions == nil {
+		return ma, errNoScorecardYmlFile
```

Comment:
Is this error (and `if` block) necessary? `if ma.Exemptions == nil`, then `ma == MaintainerAnnotations{}`.

The error is currently used below to just set `ma = MaintainerAnnotations{}` again.

Reply:
Even if the idea is to have more fields in the .scorecard.yml file in the future (i.e. the path to the release workflow), I believe that the lack of a maintainer annotations block in the file will still always lead to `ma = MaintainerAnnotations{}`?

Reply:
Sorry, I did not understand the issue. The block is necessary to set the proper error and not return this Maintainers Annotations as success case. Once we add more configurations to this file we might want to change this behavior, for example, checking if other "required" fields are missing or not.

Reply:
> The block is necessary to set the proper error and not return this Maintainers Annotations as success case. 

I just don't understand why that is necessary.

The `ma` that is returned here is basically just `MaintainerAnnotations{}`. If you omit this `if` block, you're still going to return the same `MaintainerAnnotations{}` (now without an error).

When this function's output is used by `GetMaintainersAnnotation`, it checks to see if the `errNoScorecardYmlFile` error was raised:
- If it was raised, then it returns `MaintainerAnnotations{}` (which is identical to the `ma` returned by this function.
- If it wasn't raised, then it returns the `ma` returned by this function (which in this case is identical to `MaintainerAnnotations{}`)

So, with this `if` block and associated logic in `GetMaintainerAnnotations`, you ensure that a missing/empty YML file leads to `MaintainerAnnotations{}` being returned in the end.

But if you remove this `if` block (and logic in `GetMaintainerAnnotations`), you already get `MaintainerAnnotations{}`.

Reply:
Got it, maybe it's indeed unnecessary. Will review after we make a decision on the naming.